libimage/filters.go

```diff
@@ -268,8 +268,15 @@ func filterManifest(ctx context.Context, value bool) filterFunc {
 }
 
 // filterReferences creates a reference filter for matching the specified value.
-func filterReferences(value string) filterFunc {
+func filterReferences(r *Runtime, value string) filterFunc {
+	lookedUp, _, _ := r.LookupImage(value, nil)
 	return func(img *Image) (bool, error) {
+		if lookedUp != nil {
+			if lookedUp.ID() == img.ID() {
+				return true, nil
+			}
+		}
+
```

Comment:
The code below seems to already try to handle digests? Can you clarify why this is not working and can the code below be removed to not cover the digest case?

Reply:
I don't think the code below can be removed.  The `NamesReferences` usually do not include the digest but only the tag of the image.  There are also multiple notations for the same image such as 
* alpine:latest
* alpine:latest@sha
* alpine@sha

Almost all cases are covered by `LookupImage`, so I used it to keep things simple.  However, I am not sure it will cover all cases for a reference filter which is using `path.Match`, so I refrained from altering the below logic.

Does this explanation suffice?  Shall I drop this as a code comment?

Reply:
AFAIK there is major corner case in how we lookup images: https://github.com/containers/common/issues/1248#issuecomment-1334117373
I assume this change here would match images only were the digest matches but not where the domain/patch is correct.
Using your podman PR, I see this:

```
$ bin/podman images --digests 
REPOSITORY                                  TAG         DIGEST                                                                   IMAGE ID      CREATED       SIZE
docker.io/library/alpine                    latest      sha256:02bb6f428431fbc2809c5d1b41eab5a68350194fb508869a33cb1af4444c9b11  5e2b554c1c45  5 weeks ago   7.62 MB
quay.io/libpod/testimage                    20221018    sha256:c782d03c968791b10300fb15478f7555be560329c5182ea27cba5fc299c98911  f5a99120db64  7 months ago  12.7 MB

# Now filter name from one with digest from the other:
$ bin/podman images docker.io/library/alpine@sha256:c782d03c968791b10300fb15478f7555be560329c5182ea27cba5fc299c98911
REPOSITORY                TAG         IMAGE ID      CREATED       SIZE
quay.io/libpod/testimage  20221018    f5a99120db64  7 months ago  12.7 MB
```

This seems broken, I assume the filter logic to actually use the name and not only the digest.

Reply:
Nice catch!  The code here, I think, does the right thing but image look up is broken, see:
```
podman (vendor-common) $ podman images --digests
REPOSITORY                         TAG         DIGEST                                                                   IMAGE ID      CREATED      SIZE
registry.fedoraproject.org/fedora  latest      sha256:add2f2878b383c6527a4f692c18f4aca039e7f24f5b24df1a6808f2a95280bbd  ded636d6da3e  9 days ago   196 MB
registry.access.redhat.com/ubi9    latest      sha256:8614ce95268b970880a1eca97dddfce5154fab35418d839c5f75012cccaca0d9  e7236a3e070f  6 weeks ago  217 MB
podman (vendor-common) $ podman run --rm ubi9@sha256:add2f2878b383c6527a4f692c18f4aca039e7f24f5b24df1a6808f2a95280bbd cat /etc/os-release
NAME="Fedora Linux"
```

If the image doesn't have said digest, we'll go ahead and only look up with the digest.  AFAIR, that's an inheritance from sins of the past (early podman days).

Here it becomes obvious how bad this look up is.

Reply:
I'll see what I can do and add a fix for that as well.

Reply:
Yes that is what I pointed out in the issue link. I agree if the lookup works correct and uses the name the code here should behave correctly as well.

Reply:
Are you Ok with merging this as is and revive https://github.com/containers/common/issues/1248?

That would be my preference to fix https://github.com/containers/podman/issues/18445 as much as possible.  I think that the crazy digest matching may be a longer journey since Podman and Buildah have some dependencies on the behavior.

Reply:
Yes, it properly makes more sense to have it be consistent with `podman run ... @sha:xxx`.